app.rb

```diff
@@ -175,6 +175,50 @@ def respond_with(format, category, responses)
     end
   end
 
+  def form_xml_for_settings(tag, values)
+    Nokogiri::XML::Builder.new do |xml|
+      xml.root do
+        values.each { |value| xml.send(tag, name: value) }
+      end
+    end.doc.css(tag)
+  end
+
+  get '/v1/providers.?:ext?' do
+    validate_params_ext
+
+    responses = form_xml_for_settings('provider', settings.providers)
+
+    respond_with params[:ext], :providers, responses
+  end
+
+  get '/v1/:provider/servers/types.?:ext?' do
+    validate_params_ext
+    validate_params_provider
+
+    responses = form_xml_for_settings('type', settings.server_types)
+
+    respond_with params[:ext], :types, responses
+  end
+
+  get '/v1/images/states.?:ext?' do
+    validate_params_ext
+
+    responses = form_xml_for_settings('state', settings.image_states)
+
+    respond_with params[:ext], :states, responses
+  end
+
+  get '/v1/:provider/regions.?:ext?' do
+    validate_params_ext
+    validate_params_provider
+
+    responses = form_xml_for_settings(
+      'region', settings.regions[params[:provider]]
+    )
+
+    respond_with params[:ext], :regions, responses
```

Comment:
This has to be revisited considering https://bugzilla.suse.com/show_bug.cgi?id=1169180. In Azure there are multiple names that need to get mapped, thus the simple loop over the data may not be sufficient. We have to have a conversation about this. On the other hand we may not care to expose all possible values through this API, in which case the current implementation is sufficient.

Reply:
Yes, we need to extend the 'region list' design that's currently in Azure into the other providers, and extend it to have a canonical region name, with acceptable aliases. This is going to be a pretty big rework project :/